src/ng/directive/ngEventDirs.js

```diff
@@ -51,32 +51,36 @@ forEach(
   function(eventName) {
     var directiveName = directiveNormalize('ng-' + eventName);
     ngEventDirectives[directiveName] = ['$parse', '$rootScope', function($parse, $rootScope) {
-      return {
-        restrict: 'A',
-        compile: function($element, attr) {
-          // NOTE:
-          // We expose the powerful `$event` object on the scope that provides access to the Window,
-          // etc. This is OK, because expressions are not sandboxed any more (and the expression
-          // sandbox was never meant to be a security feature anyway).
-          var fn = $parse(attr[directiveName]);
-          return function ngEventHandler(scope, element) {
-            element.on(eventName, function(event) {
-              var callback = function() {
-                fn(scope, {$event: event});
-              };
-              if (forceAsyncEvents[eventName] && $rootScope.$$phase) {
-                scope.$evalAsync(callback);
-              } else {
-                scope.$apply(callback);
-              }
-            });
-          };
-        }
-      };
+      return createEventDirective($parse, $rootScope, directiveName, eventName, forceAsyncEvents[eventName]);
     }];
   }
 );
 
+function createEventDirective($parse, $rootScope, attrName, eventName, forceAsync) {
```

Comment:
We could have a `value` param on this function which is used rather than `attr[attrName]` if it is defined.
That way you can avoid adding the value to the `attrs` map in the case of `ng-on-`, no?

Reply:
I wanted to avoid making changes to the existing event directives, and I thought making `ng-on-*` the same would also be good. In all those cases the `attr[attrName]` is not read until compile time...